text/0033-crash-recovery.md

```diff
@@ -0,0 +1,108 @@
+    - The hash of the latest unit
+    - The list of hashes of endorsed units
+
+This data should be keyed by the consensus instance ID.
+
+In order to facilitate removing data for old eras (ones that are older than `UNBONDING_PERIOD`), we should also store a map of era ID → instance ID - as instance IDs themselves carry no information about how old the era is. We can, however, identify the era IDs of old eras, and using this map, we could also find the entries to be removed.
+
+Any sending of a unit / an endorsement should only happen after the storage request returns a success response.
+
+- Whenever a new era is started, it should first try to read the entry with latest hashes from storage by submitting a read request.
+- If the entry doesn't exist, proceed the same way it is done currently.
+- If the entry exists:
+    - Pass all the data to the `ActiveValidator` instance.
+    - Only create new units if the unit with the latest hash is known to the consensus state.
+    - Only create new endorsements if all the previous endorsements are known to the consensus state.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+Every creation of a unit or an endorsement will now involve storage - however, this seems unavoidable in that we need to store the knowledge of our activity in consensus in some permanent way.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### An alternative hash saving scheme
+
+In order to prevent situations in which we get stuck with a unit hash that no other validator has heard of, we could use a solution like the following:
+
+1. Let's denote n-th unit created by us by U[n].
+2. Whenever we create a unit, we _append_ its hash to the list of hashes of the units we created.
+3. Whenever we receive a unit citing U[n] from another validators whose weights add up to more than our FTT, we remove the hashes of U[n-1] and older units from the list.
+4. When we rejoin an era and try to sync the state, request the latest unit from the list from other validators.
+5. If no validator has this unit, remove it from the list and try with the next latest one.
+
+Point 2 makes sure that at least one honest validator has received our unit before we remove its ancestors from our list. This way, we should always have at least one hash that corresponds to a unit received by an honest validator.
+
+Point 5 makes sure that we don't try starting from an older unit unless no validator seems able to send it to us.
+
+However, this scheme still has a weakness. It is theoretically possible that we will remove all the older hashes from the list when just a single honest validator confirms its reception, and that this validator will crash right afterwards. In such a case we will still be stuck with just hashes of units that no active honest validator has heard of and that we won't be able to sync. The risk should be significantly lower, though.
```

Comment:
Another drawback is that the only node with our latest unit could be malicious. They could withhold it from us until we drop it from the list and create an equivocating unit. Then they could publish it and have proof of us equivocating.

Reply:
Right. I convinced myself at some point that this couldn't happen, but now I think you're right and actually it could. So good point :+1: